src/contracts/exchange/NiftyswapExchange20.sol

```diff
@@ -0,0 +1,812 @@
+      uint256 currencyReserve = currencyReserves[id];
+
+      // Calculate amount to withdraw for currency and Token _id
+      uint256 currencyAmount = amountPool.mul(currencyReserve) / totalLiquidity;
+      uint256 tokenAmount = amountPool.mul(tokenReserve) / totalLiquidity;
+
+      // Verify if amounts to withdraw respect minimums specified
+      require(currencyAmount >= _minCurrency[i], "NiftyswapExchange#_removeLiquidity: INSUFFICIENT_CURRENCY_AMOUNT");
+      require(tokenAmount >= _minTokens[i], "NiftyswapExchange#_removeLiquidity: INSUFFICIENT_TOKENS");
+
+      // Update total liquidity pool token supply of Token _id
+      totalSupplies[id] = totalLiquidity.sub(amountPool);
+
+      // Update currency reserve size for Token id
+      currencyReserves[id] = currencyReserve.sub(currencyAmount);
+
+      // Update totalCurrency and tokenAmounts
+      totalCurrency = totalCurrency.add(currencyAmount);
+      tokenAmounts[i] = tokenAmount;
+      currencyAmounts[i] = currencyAmount;
+    }
+
+    // Burn liquidity pool tokens for offchain supplies
+    _batchBurn(address(this), _tokenIds, _poolTokenAmounts);
+
+    // Transfer total currency and all Tokens ids
+    currency.transfer(_provider, totalCurrency);
+    token.safeBatchTransferFrom(address(this), _provider, _tokenIds, tokenAmounts, "");
+
+    // Emit event
+    emit LiquidityRemoved(_provider, _tokenIds, tokenAmounts, currencyAmounts);
+  }
+
+  /***********************************|
+  |     Receiver Methods Handler      |
+  |__________________________________*/
+
+  // Method signatures for onReceive control logic
+
+  // bytes4(keccak256(
+  //   "_tokenToCurrency(uint256[],uint256[],uint256,uint256,address)"
+  // ));
+  bytes4 internal constant SELLTOKENS_SIG = 0xdb08ec97;
+
+  //  bytes4(keccak256(
+  //   "_addLiquidity(address,uint256[],uint256[],uint256[],uint256)"
+  // ));
+  bytes4 internal constant ADDLIQUIDITY_SIG = 0x82da2b73;
+
+  // bytes4(keccak256(
+  //    "_removeLiquidity(address,uint256[],uint256[],uint256[],uint256[],uint256)"
+  // ));
+  bytes4 internal constant REMOVELIQUIDITY_SIG = 0x5c0bf259;
+
+  // bytes4(keccak256(
+  //   "DepositTokens()"
+  // ));
+  bytes4 internal constant DEPOSIT_SIG = 0xc8c323f9;
+
+  /***********************************|
+  |           Buying Tokens           |
+  |__________________________________*/
+
+  /**
+   * @dev Purchase tokens with currency.
+   */
+  function buyTokens(
+    uint256[] memory _tokenIds,
+    uint256[] memory _tokensBoughtAmounts,
+    uint256 _maxCurrency,
+    uint256 _deadline,
+    address _recipient
+  )
+  override external returns (uint256[] memory)
+  {
+    require(_deadline >= block.timestamp, "NiftyswapExchange#buyTokens: DEADLINE_EXCEEDED");
+    require(_tokenIds.length > 0, "NiftyswapExchange#buyTokens: INVALID_CURRENCY_IDS_AMOUNT");
+
+    // Transfer the tokens for purchase
+    currency.transferFrom(msg.sender, address(this), _maxCurrency);
```

Comment:
Should require that this returns true since not all ERC-20 transfer implementation will throw on error.